cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1155 @@
+			vid := newVersionedID(partition.ID, tag)
+			vid.target = unsafe.Pointer(tbInfo)
+			s.partitions.ReplaceOrInsert(vid)
+			if ineligible {
+				s.ineligibleTables.ReplaceOrInsert(newVersionedID(partition.ID, tag))
+			}
+		}
+	}
+}
+
+// updatePartition updates partition info for `tbInfo`.
+func (s *Snapshot) updatePartition(tbInfo *model.TableInfo, currentTs uint64) error {
+	oldTbInfo, ok := s.PhysicalTableByID(tbInfo.ID)
+	if !ok {
+		return cerror.ErrSnapshotTableNotFound.GenWithStackByArgs(tbInfo.ID)
+	}
+	oldPi := oldTbInfo.GetPartitionInfo()
+	if oldPi == nil {
+		return cerror.ErrSnapshotTableNotFound.GenWithStack("table %d is not a partition table", tbInfo.ID)
+	}
+	newPi := tbInfo.GetPartitionInfo()
+	if newPi == nil {
+		return cerror.ErrSnapshotTableNotFound.GenWithStack("table %d is not a partition table", tbInfo.ID)
+	}
+
+	tag := negative(currentTs)
+	vid := newVersionedID(tbInfo.ID, tag)
+	vid.target = unsafe.Pointer(tbInfo.Clone())
+	s.tables.ReplaceOrInsert(vid)
+	ineligible := !tbInfo.IsEligible(s.forceReplicate)
+	if ineligible {
+		s.ineligibleTables.ReplaceOrInsert(newVersionedID(tbInfo.ID, tag))
+	}
+	for _, partition := range oldPi.Definitions {
+		s.partitions.ReplaceOrInsert(newVersionedID(partition.ID, tag))
+	}
+	for _, partition := range newPi.Definitions {
+		vid := newVersionedID(partition.ID, tag)
+		vid.target = unsafe.Pointer(tbInfo)
+		s.partitions.ReplaceOrInsert(vid)
+		if ineligible {
+			s.ineligibleTables.ReplaceOrInsert(newVersionedID(partition.ID, tag))
+		}
+	}
+	s.currentTs = currentTs
+	// TODO: is it necessary to print changes detailly?
+	log.Debug("adjust partition success",
+		zap.String("schema", tbInfo.TableName.Schema),
+		zap.String("table", tbInfo.TableName.Table))
+	return nil
+}
+
+func (s *Snapshot) renameTables(job *timodel.Job, currentTs uint64) error {
+	var oldSchemaIDs, newSchemaIDs, oldTableIDs []int64
+	var newTableNames, oldSchemaNames []*timodel.CIStr
+	err := job.DecodeArgs(&oldSchemaIDs, &newSchemaIDs, &newTableNames, &oldTableIDs, &oldSchemaNames)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(job.BinlogInfo.MultipleTableInfos) < len(newTableNames) {
+		return cerror.ErrInvalidDDLJob.GenWithStackByArgs(job.ID)
+	}
+	// NOTE: should handle failures in halfway better.
+	for _, tableID := range oldTableIDs {
+		if err := s.dropTable(tableID, currentTs); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	for i, tableInfo := range job.BinlogInfo.MultipleTableInfos {
+		newSchema, ok := s.SchemaByID(newSchemaIDs[i])
+		if !ok {
+			return cerror.ErrSnapshotSchemaNotFound.GenWithStackByArgs(newSchemaIDs[i])
+		}
+		newSchemaName := newSchema.Name.L
+		tbInfo := model.WrapTableInfo(newSchemaIDs[i], newSchemaName, job.BinlogInfo.FinishedTS, tableInfo)
+		err = s.createTable(tbInfo, currentTs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// HandleDDL handles the given job.
+func (s *Snapshot) HandleDDL(job *timodel.Job) error {
```

Comment:
Is is thread-safe?

Reply:
Yes all public functions are thread-safe. I will comment it on `Snapshot`.